File: leetcode/rust-leetcode/src/2556.disconnect-path-in-a-binary-matrix-by-at-most-one-flip.rs
impl Solution {
    pub fn is_possible_to_cut_path(grid: Vec<Vec<i32>>) -> bool {
        let (m, n) = (grid.len(), grid[0].len());
        fn check_path(grid: &mut Vec<Vec<i32>>, i: usize, j: usize, m: usize, n: usize) -> bool {
            if i == m && j == n {
                return true;
            }
            if i > m || j > n {
                return false;
            }
            if grid[i][j] == 0 {
                return false;
            }
            grid[i][j] = 0;
            check_path(grid, i+1, j, m, n) || check_path(grid, i, j+1, m, n)
        }
        let mut grid = grid;
        if !check_path(&mut grid, 0, 0, m-1, n-1) {
            return true;
        }
        grid[0][0] = 1;
        !check_path(&mut grid, 0, 0, m-1, n-1)
    }
}

File: leetcode/python/2556.disconnect-path-in-a-binary-matrix-by-at-most-one-flip.py
class Solution:
    def isPossibleToCutPath(self, grid: List[List[int]]) -> bool:
        m, n = len(grid)-1, len(grid[0])-1

        def pathExists(i: int, j: int) -> bool:
            if i == m and j == n:
                return True
            if i > m or j > n:
                return False
            if grid[i][j] == 0:
                return False
            grid[i][j] = 0
            return pathExists(i + 1, j) or pathExists(i, j+1)
        
        if not pathExists(0, 0):
            return True

        grid[0][0] = 1

        return not pathExists(0, 0)

